sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java

```diff
@@ -0,0 +1,303 @@
+    insertTestDocuments(NB_DOCS);
+
+    String query = "{\n"
+        + "  \"query\": {\n"
+        + "  \"match\" : {\n"
+        + "    \"scientist\" : {\n"
+        + "      \"query\" : \"Einstein\",\n"
+        + "      \"type\" : \"boolean\"\n"
+        + "    }\n"
+        + "  }\n"
+        + "  }\n"
+        + "}";
+
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(connectionConfiguration)
+                              .withQuery(query));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(NB_DOCS / 10);
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NB_DOCS, false);
+    pipeline.apply(Create.of(data)).apply(ElasticsearchIO.write()
+      .withConnectionConfiguration(connectionConfiguration));
+    pipeline.run();
+
+    // upgrade
+    node.client().admin().indices().upgrade(new UpgradeRequest()).actionGet();
+
+    SearchResponse response = node.client().prepareSearch().execute().actionGet(5000);
+    assertEquals(NB_DOCS, response.getHits().getTotalHits());
+
+    QueryBuilder queryBuilder = QueryBuilders.queryStringQuery("Einstein").field("scientist");
+    response = node.client().prepareSearch().setQuery(queryBuilder).execute().actionGet();
+    assertEquals(NB_DOCS / 10, response.getHits().getTotalHits());
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWriteWithErrors() throws Exception {
+    TestPipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NB_DOCS, true);
+    pipeline.apply(Create.of(data)).apply(ElasticsearchIO.write().
+      withConnectionConfiguration(connectionConfiguration));
+    try {
+      pipeline.run();
+      fail("Exception should have been thrown");
+    } catch (Exception e) {
+        //type of e is Pipeline$PipelineExecutionException
+
+      String message = e.getCause().getMessage();
+      assertTrue("Wrong exception thrown", e.getCause().getClass() == IOException.class);
+      assertTrue("Wrong message in exception",
+        message.matches(
+          "(?is).*Error writing to Elasticsearch, some elements could not be inserted"
+              + ".*document id.*was expecting a colon to separate field name and value"
+              + ".*document id.*was expecting a colon to separate field name and value.*"));
+    }
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWriteWithBatchSizes() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NB_DOCS, false);
+    PDone collection = pipeline.apply(Create.of(data)).apply(
+        ElasticsearchIO.write()
+            .withConnectionConfiguration(connectionConfiguration)
+            .withBatchSize(NB_DOCS / 2)
```

Comment:
-  +1 for non critical parameters (like for previous comment)

- Considering withBatchSize*() tests: I thought I could force the number of bundles but I realize that this is runner specific, so the code above is pointless. In fact withBatchSize*() are more with**Max**BatchSize() because the runner may split data before specified batchSize is reached. So it might be misleading for users to have these parameters in the API. They might expect that they behave the same as corresponding ElasticSearch bulk parameters (create fixed size batches). @jkff, @jbonofre any thoughts? Should they be kept? If yes, how should they be properly tested?